### DISCUSSION TITLE
input.http error

### DISCUSSION THREAD

**Discussion by 918483** [savonet/liquidsoap#2354] [User Support]
Why the error below occurs in the 2.0.4 update

relay = input.http(id="relay", restart=true, restart_on_error=false, buffer=0.2, max=999., url)
Error 5: this value has no method http

**Comment by toots** [comment]
Hi,

You need to have `ocaml-ffmpeg` installed for this operator to be enabled.

**Comment by 918483** [comment]
using the command below the following error ?

cd /ocaml

git clone https://github.com/savonet/ocaml-ffmpeg.git
cd ocaml-ffmpeg
opam pin add .

---------------------ERROR---------------------------

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of ffmpeg-avutil.1.1.4 failed at "dune build -p ffmpeg-avutil -j 1 @install".

#=== ERROR while compiling ffmpeg-avutil.1.1.4 ================================#

context 2.1.2 | linux/x86_64 | ocaml-base-compiler.4.12.0 | pinned(git+file:///ocaml/ocaml-ffmpeg#main#6ca1aaf1)
path ~/.opam/4.12.0/.opam-switch/build/ffmpeg-avutil.1.1.4
command ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ffmpeg-avutil -j 1 @install
exit-code 1
env-file ~/.opam/log/ffmpeg-avutil-31814-6adf47.env
output-file ~/.opam/log/ffmpeg-avutil-31814-6adf47.out
output
[...]
avutil_stubs.c:1620:13: error: request for member ‘sw_format’ in something not a structure or union
frames_ctx->sw_format = PixelFormat_val(_src_pixel_format);
^
avutil_stubs.c:1621:13: error: request for member ‘width’ in something not a structure or union
frames_ctx->width = Int_val(_width);
^
avutil_stubs.c:1622:13: error: request for member ‘height’ in something not a structure or union
frames_ctx->height = Int_val(_height);
^
avutil_stubs.c:1625:3: warning: implicit declaration of function ‘av_hwframe_ctx_init’ [-Wimplicit-function-declaration]
ret = av_hwframe_ctx_init(hw_frames_ref);
^
<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build ffmpeg-avutil 1.1.4
└─

**Reply by toots** [reply]
I'm not sure but you shouldn't need to `pin` it, the latest release version, `1.1.3` should be sufficient.

**Reply by 918483** [reply]
Could you tell me how to do it?

**Reply by 918483** [reply]
Only this part is missing to relay the code

**Reply by toots** [reply]
`opam install ffmpeg`

**Reply by 918483** [reply]
The following actions failed
│ λ build ffmpeg-avutil 1.1.3
└─
┌─ The following changes have been performed (the rest was aborted)
│ ∗ install conf-ffmpeg       1
│ ∗ install conf-pkg-config   2
│ ∗ install csexp             1.5.1
│ ∗ install dune              3.1.1
│ ∗ install dune-configurator 3.1.1

**Reply by toots** [reply]
Do you have have more logs?

**Reply by 918483** [reply]
```
root@vps790319 [/]# opam install ffmpeg
[WARNING] Running as root is not recommended
The following actions will be performed:
  ∗ install ffmpeg-avutil     1.1.3 [required by ffmpeg]
  ∗ install ffmpeg-swscale    1.1.3 [required by ffmpeg]
  ∗ install ffmpeg-avfilter   1.1.3 [required by ffmpeg]
  ∗ install ffmpeg-avcodec    1.1.3 [required by ffmpeg]
  ∗ install ffmpeg-swresample 1.1.3 [required by ffmpeg]
  ∗ install ffmpeg-av         1.1.3 [required by ffmpeg]
  ∗ install ffmpeg-avdevice   1.1.3 [required by ffmpeg]
  ∗ install ffmpeg            1.1.3
===== ∗ 8 =====
Do you want to continue? [Y/n] y

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⬇ retrieved ffmpeg.1.1.3  (cached)
⬇ retrieved ffmpeg-av.1.1.3  (cached)
⬇ retrieved ffmpeg-avcodec.1.1.3  (cached)
⬇ retrieved ffmpeg-avdevice.1.1.3  (cached)
⬇ retrieved ffmpeg-avfilter.1.1.3  (cached)
⬇ retrieved ffmpeg-avutil.1.1.3  (cached)
⬇ retrieved ffmpeg-swresample.1.1.3  (cached)
⬇ retrieved ffmpeg-swscale.1.1.3  (cached)
[ERROR] The compilation of ffmpeg-avutil.1.1.3 failed at "dune build -p ffmpeg-avutil -j 1 @install".

#=== ERROR while compiling ffmpeg-avutil.1.1.3 ================================#
# context     2.1.2 | linux/x86_64 | ocaml-base-compiler.4.14.0 | https://opam.ocaml.org#6ddead68
# path        ~/.opam/4.14.0/.opam-switch/build/ffmpeg-avutil.1.1.3
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ffmpeg-avutil -j 1 @install
# exit-code   1
# env-file    ~/.opam/log/ffmpeg-avutil-10378-5ee527.env
# output-file ~/.opam/log/ffmpeg-avutil-10378-5ee527.out
### output ###
# [...]
# avutil_stubs.c:1621:13: error: request for member ‘sw_format’ in something not a structure or union
#    frames_ctx->sw_format = PixelFormat_val(_src_pixel_format);
#              ^
# avutil_stubs.c:1622:13: error: request for member ‘width’ in something not a structure or union
#    frames_ctx->width = Int_val(_width);
#              ^
# avutil_stubs.c:1623:13: error: request for member ‘height’ in something not a structure or union
#    frames_ctx->height = Int_val(_height);
#              ^
# avutil_stubs.c:1626:3: warning: implicit declaration of function ‘av_hwframe_ctx_init’ [-Wimplicit-function-declaration]
#    ret = av_hwframe_ctx_init(hw_frames_ref);
#    ^

<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build ffmpeg-avutil 1.1.3
└─
╶─ No changes have been performed
# Run eval $(opam env) to update the current shell environment
```

**Reply by 918483** [reply]
It doesn't get out of this error at all

**Reply by toots** [reply]
This seems to indicate that your version of `ffmpeg` is too old.

**Reply by 918483** [reply]
cd /
wget https://www.ffmpeg.org/releases/ffmpeg-5.0.tar.gz
tar -zxvf ffmpeg-5.0.tar.gz
cd ffmpeg-5.0
./configure --enable-nonfree --enable-openssl --disable-yasm --enable-libmp3lame --enable-pic --enable-gpl --enable-shared --enable-avutil --enable-decoder=aac --enable-filter=aformat --enable-filter=volume --enable-filter=aresample && make && make install && ldconfig

**Reply by toots** [reply]
Are you sure that you have removed all other versions? What's the output of:
```
% pkg-config --libs libavutil
```

**Reply by 918483** [reply]
root@vps790319 [/]# pkg-config --libs libavutil
-lavutil
root@vps790319 [/]#

**Comment by toots** [comment]
Try this:
```
% pkg-config --variable=libdir libavutil
```

**Reply by 918483** [reply]
root@vps790319 [/]# pkg-config --variable=libdir libavutil
/usr/lib64
root@vps790319 [/]#

**Reply by toots** [reply]
Yeah it looks like you already have ffmpeg installed in your system. You should either:
* Remove the system ffmpeg
* Make sure that `PKG_CONFIG_PATH` points to the locally compiled ffmpeg.

**Reply by 918483** [reply]
Could you tell me how to install it correctly? I already removed ffmpeg so I didn't understand the PKG_CONFIG_PATH part

**Reply by 918483** [reply]
I've already installed more than 5x the system and always the same error would have another way to install I'm using opam

ERROR while compiling ffmpeg-avutil.1.1.3
ERROR while compiling ocurl.0.9.2

Both errors in 2.0.4 compilation

**Reply by 918483** [reply]
[erros.zip](https://github.com/savonet/liquidsoap/files/8585458/erros.zip)

**Reply by 918483** [reply]
could you tell me how to compile it locally?
because I have already formatted the machine more than 8x and every time I install the error even without installing ffmpeg if you can give me a step by step installation I appreciate the help.


https://www.elephdev.com/index.php/cphp/317.html?ref=addtabs&lang=en

**Reply by toots** [reply]
You described it here: https://github.com/savonet/liquidsoap/discussions/2354#discussioncomment-2650070

As far as I can see you need to do the steps in that comment and then make sure to remove centos ffmpeg packages:
```
yum remove ffmpeg ffmpeg-devel
```

**Reply by 918483** [reply]
![Screenshot_1](https://user-images.githubusercontent.com/20747172/165855783-02f1dbfc-f143-4648-a480-c26426e5b2ec.png)

**Reply by 918483** [reply]
does not exist installed

**Reply by 918483** [reply]
![Screenshot_2](https://user-images.githubusercontent.com/20747172/165856180-8f249397-6d4c-4da1-b511-dde78a79f490.png)

**Reply by toots** [reply]
If you have installed `ffmpeg` manually you should skip the depext and avoid installing it again via the system packages. So, no `opam depext ffmpeg`

**Reply by 918483** [reply]
![Screenshot_1](https://user-images.githubusercontent.com/20747172/165856698-e5290339-cc25-4ae3-9fb9-47205c2f20fa.png)

**Reply by 918483** [reply]
now it's another mistake @toots

**Reply by 918483** [reply]
used opam install -y --assume-depext ffmpeg

**Reply by toots** [reply]
It depends on your `opam` version, for old version just do:
```
opam install -y ffmpeg
```
And no call to `opam depext` before.

**Reply by toots** [reply]
Ha I see. That means that `pkg-config` cannot find the ffmpeg installed locally.

* Have you repeated the steps at: https://github.com/savonet/liquidsoap/discussions/2354#discussioncomment-2650070
* Do you know where ffmpeg was installed?

If so, you should point `PKG_CONFIG_PATH` to this location:
```
export PKG_CONFIG_PATH=/path/to/ffmpeg/install/lib/pkg-config
opam install -y ffmpeg
```

As a side note, I would suggest trying to install on a more recent OS. The problem is bound to occur for other dependencies and if you are having trouble installing things manually, switching to a more recent system will make the whole install experience much better.

**Reply by 918483** [reply]
I'll go down how to install

sudo yum install epel-release -y
yum repolist -y
yum install wget bubblewrap -y
sudo yum makecache
yum groupinstall "Development Tools"
yum update -y

bash -c "sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)"
opam init
eval $(opam env)
opam switch create 4.14.0
eval $(opam env --switch=4.14.0)
which ocaml
ocaml -version

yum install faad2-devel fdk-aac-devel flac-devel lame-devel libcurl-devel libmad-devel libogg-devel libsamplerate-devel libtheora-devel libvorbis-devel openssl-devel opus-devel pcre-devel speex-devel taglib-devel bzip2 telnet telnet-server

rpm -Uvh  https://download1.rpmfusion.org/free/el/rpmfusion-free-release-7.noarch.rpm
yum install -y ffmpeg-devel

opam -y depext taglib mad lame vorbis cry flac fdkaac faad opus vorbis theora speex ogg xmlplaylist duppy ffmpeg ocurl samplerate liquidsoap
opam -y install taglib mad lame vorbis cry flac fdkaac faad opus vorbis theora speex ogg xmlplaylist duppy ffmpeg ocurl samplerate liquidsoap


Could you tell me what is wrong?

**Reply by toots** [reply]
Again, I suggest to use a more recent OS. I'm happy to help but I am not a CentOS user or administrator. Your problem is that the `ffmpeg` (and `ffmpeg-devel`) packages installed via `yum` on this OS are too old and won't work with the version of `ocaml-ffmpeg` that you are trying to install.

The steps, conceptually are:
* Do not install `ffmpeg` and `ffmpeg-devel` via `yum`. That includes:
  * Removing `ffmpeg` and `ffmpeg-devel` from the `yum install` commands
  * Removing `ffmpeg` from the `opam depext` command
 * Install a locally build `ffmpeg` as described in your own message above.
 * Make sure that `pkg-config` knows where to find `ffmpeg` (try: `pkg-config --libs libavutil`).
 * Install everything via `opam install`

**Reply by 918483** [reply]
I understand it worked, but if I change to centos 8 would it improve?

**Reply by 918483** [reply]
How do I change the function below to ffmpeg instead of input.http

funcao_relay = input.http(buffer = 5.00, max = 10.00, id="iniciar", autostart=false, "http://01.painelstreaming.com.br:7062")
funcao_relay = map_metadata(relay_title,funcao_relay)
fallback(track_sensitive = false, [funcao_relay,radio])

**Reply by 918483** [reply]
relay stream
